mistralrs-core/src/cuda/nonzero_bitwise.cu

```diff
@@ -0,0 +1,205 @@
+template <typename T>
+void count_nonzero(const T *d_in, const uint32_t N, uint32_t *h_out) {
+  cub::TransformInputIterator<bool, NonZeroOp<T>, const T *> itr(
+      d_in, NonZeroOp<T>());
+  size_t temp_storage_bytes = 0;
+  size_t *d_num_nonzero;
+  cudaMalloc((void **)&d_num_nonzero, sizeof(uint32_t));
+  cub::DeviceReduce::Sum(nullptr, temp_storage_bytes, itr, d_num_nonzero, N);
+  void **d_temp_storage;
+  cudaMalloc(&d_temp_storage, temp_storage_bytes);
+  cub::DeviceReduce::Sum(d_temp_storage, temp_storage_bytes, itr, d_num_nonzero,
+                         N);
+  cudaMemcpy(h_out, d_num_nonzero, sizeof(uint32_t), cudaMemcpyDeviceToHost);
+  cudaFree(d_num_nonzero);
+  cudaFree(d_temp_storage);
+}
+
+#define COUNT_NONZERO_OP(TYPENAME, RUST_NAME)                                  \
+  extern "C" uint32_t count_nonzero_##RUST_NAME(const TYPENAME *d_in,          \
+                                                uint32_t N) {                  \
+    uint32_t result;                                                           \
+    count_nonzero(d_in, N, &result);                                           \
+    return result;                                                             \
+  }
+
+//#if __CUDA_ARCH__ >= 800
+COUNT_NONZERO_OP(__nv_bfloat16, bf16)
+//#endif
+
+//#if __CUDA_ARCH__ >= 530
+COUNT_NONZERO_OP(__half, f16)
+//#endif
+
+COUNT_NONZERO_OP(float, f32)
+COUNT_NONZERO_OP(double, f64)
+COUNT_NONZERO_OP(uint8_t, u8)
+COUNT_NONZERO_OP(uint32_t, u32)
+COUNT_NONZERO_OP(int64_t, i64)
+
+__global__ void transform_indices(const uint32_t *temp_indices,
+                                  const uint32_t num_nonzero,
+                                  const uint32_t *dims, const uint32_t num_dims,
+                                  uint32_t *d_out) {
+  const int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx < num_nonzero) {
+    int temp_index = temp_indices[idx];
+    for (int i = num_dims - 1; i >= 0; i--) {
+      d_out[idx * num_dims + i] = temp_index % dims[i];
+      temp_index /= dims[i];
+    }
+  }
+}
+
+// get the indices of non-zero elements in an array
+template <typename T>
+void nonzero(const T *d_in, const uint32_t N, const uint32_t num_nonzero,
+             const uint32_t *dims, const uint32_t num_dims, uint32_t *d_out) {
+  cub::TransformInputIterator<bool, NonZeroOp<T>, const T *> itr(
+      d_in, NonZeroOp<T>());
+  cub::CountingInputIterator<uint32_t> counting_itr(0);
+  uint32_t *out_temp;
+  uint32_t *num_selected_out;
+  cudaMalloc((void **)&out_temp, num_nonzero * sizeof(uint32_t));
+  cudaMalloc((void **)&num_selected_out, sizeof(uint32_t));
+  size_t temp_storage_bytes = 0;
+  cub::DeviceSelect::Flagged(nullptr, temp_storage_bytes, counting_itr, itr,
+                             out_temp, num_selected_out, N);
+  void **d_temp_storage;
+  cudaMalloc(&d_temp_storage, temp_storage_bytes);
+  cub::DeviceSelect::Flagged(d_temp_storage, temp_storage_bytes, counting_itr,
```

Comment:
Will usage of `cub` require additions to the installation docs?

Reply:
cub is shipped with nvidia cuda toolkit, so any environment with Cuda should work fine

Reply:
Ok, great, thanks for clarifying!